File: python/tvm/meta_schedule/measure.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""
Taken and modified from Ansor.

Distributed measurement infrastructure to measure the runtime costs of tensor programs.

These functions are responsible for building the tvm module, uploading it to
remote devices, recording the running time costs, and checking the correctness of the output.

We separate the measurement into two steps: build and run.
A builder builds the executable binary files and a runner runs the binary files to
get the measurement results. The flow of data structures is

                `ProgramBuilder`                 `ProgramRunner`
`MeasureInput` -----------------> `BuildResult` ----------------> `MeasureResult`

We implement these in python to utilize python's multiprocessing and error handling.
"""
from __future__ import annotations

import os
from typing import TYPE_CHECKING, Callable, List, Optional, Union

from tvm._ffi import register_object
from tvm.runtime import Object

from ..runtime import NDArray, TVMContext, ndarray
from . import _ffi_api
from .measure_record import BuildResult, MeasureInput, MeasureResult
from .schedule import Schedule
from .utils import check_remote_servers, cpu_count

if TYPE_CHECKING:
    import numpy as np

    from ..target import Target
    from ..tir import PrimFunc


########## ProgramBuilder ##########


@register_object("meta_schedule.ProgramBuilder")
class ProgramBuilder(Object):
    """ The base class of ProgramBuilders. """

    n_parallel: int
    timeout: int

    @staticmethod
    def create(name: str) -> "ProgramBuilder":
        if name == "local":
            return LocalBuilder()
        raise ValueError("Unknown name of program builder: " + name)

    def build(
        self,
        measure_inputs: List[MeasureInput],
        verbose: int = 1,
    ) -> List[BuildResult]:
        """Build programs and return results.

        Parameters
        ----------
        measure_inputs : List[MeasureInput]
            A List of MeasureInput.
        verbose: int = 1
            Verbosity level. 0 for silent, 1 to output information during program building.

        Returns
        -------
        res : List[BuildResult]
        """
        return _ffi_api.ProgramBuilderBuild(  # pylint: disable=no-member
            self,
            measure_inputs,
            verbose,
        )


########## ProgramRunner ##########


@register_object("meta_schedule.ProgramRunner")
class ProgramRunner(Object):
    """ The base class of ProgramRunners. """

    timeout: int
    number: int
    repeat: int
    min_repeat_ms: int
    cooldown_interval: float
    enable_cpu_cache_flush: bool

    @staticmethod
    def create(name: str) -> "ProgramRunner":
        if name == "rpc":
            return RPCRunner()
        raise ValueError("Unknown name of program builder: " + name)

    def run(
        self,
        measure_inputs: List[MeasureInput],
        build_results: List[BuildResult],
        verbose: int = 1,
    ) -> List[MeasureResult]:
        """Run measurement and return results.

        Parameters
        ----------
        measure_inputs : List[MeasureInput]
            A List of MeasureInput.
        build_results : List[BuildResult]
            A List of BuildResult to be ran.
        verbose: int = 1
            Verbosity level. 0 for silent, 1 to output information during program running.

        Returns
        -------
        res : List[MeasureResult]
        """
        return _ffi_api.ProgramRunnerRun(  # pylint: disable=no-member
            self,
            measure_inputs,
            build_results,
            verbose,
        )


########## LocalBuilder ##########


@register_object("meta_schedule.LocalBuilder")
class LocalBuilder(ProgramBuilder):
    """LocalBuilder use local CPU cores to build programs in parallel."""

    build_func: str

    def __init__(
        self,
        timeout: int = 15,
        n_parallel: Optional[int] = None,
        build_func: str = "tar",
    ):
        if n_parallel is None:
            n_parallel = cpu_count()
        self.__init_handle_by_constructor__(
            _ffi_api.LocalBuilder,  #  pylint: disable=no-member
            timeout,
            n_parallel,
            build_func,
        )


########## RPCRunner ##########


@register_object("meta_schedule.RPCRunner")
class RPCRunner(ProgramRunner):
    """RPCRunner that uses RPC call to measures the time cost of programs on remote devices.
    Or sometime we may need to use RPC even in local running to insulate the thread environment.
    (e.g. running CUDA programs)"""

    key: str
    host: str
    port: int
    priority: int
    n_parallel: int

    def __init__(
        self,
        key: Optional[str] = None,
        host: Optional[str] = None,
        port: Optional[int] = None,
        priority: int = 1,
        n_parallel: Optional[int] = None,
        timeout: int = 10,
        number: int = 3,
        repeat: int = 1,
        min_repeat_ms: int = 40,
        cooldown_interval: float = 0.0,
        enable_cpu_cache_flush: bool = False,
        f_create_args: Optional[Callable[[TVMContext], List[NDArray]]] = None
    ):
        if key is None:
            key = os.environ.get("TVM_TRACKER_KEY", None)
            if key is None:
                raise ValueError(
                    "RPC device key is not provided. Please provide 'key' explicitly, "
                    "or set environment variable TVM_TRACKER_KEY"
                )
        if host is None:
            host = os.environ.get("TVM_TRACKER_HOST", None)
            if host is None:
                raise ValueError(
                    "RPC tracker's host address is not provided. Please provide 'host' explicitly, "
                    "or set environment variable TVM_TRACKER_HOST"
                )
        if port is None:
            port = os.environ.get("TVM_TRACKER_PORT", None)
            if port is None:
                raise ValueError(
                    "RPC tracker's host address is not provided. Please provide 'port' explicitly, "
                    "or set environment variable TVM_TRACKER_PORT"
                )
            port = int(port)

        n_servers = check_remote_servers(key, host, port, priority, timeout)

        if n_parallel is None:
            n_parallel = n_servers

        self.__init_handle_by_constructor__(
            _ffi_api.RPCRunner,  #  pylint: disable=no-member
            key,
            host,
            port,
            priority,
            n_parallel,
            timeout,
            number,
            repeat,
            min_repeat_ms,
            cooldown_interval,
            enable_cpu_cache_flush,
            f_create_args,
        )


########## MeasureCallback ##########


@register_object("meta_schedule.MeasureCallback")
class MeasureCallback(Object):
    """The base class of measurement callback functions."""


@register_object("meta_schedule.RecordToFile")
class RecordToFile(Object):
    """The base class of measurement callback functions."""

    def __init__(self):
        self.__init_handle_by_constructor__(_ffi_api.RecordToFile)  # pylint: disable=no-member


########## ProgramMeasurer ##########


@register_object("meta_schedule.ProgramMeasurer")
class ProgramMeasurer(Object):
    """Measurer that measures the time costs of tvm programs.

    Parameters
    ----------
    builder : ProgramBuilder
        The ProgramBuilder to build each program
    runner : ProgramRunner
        The ProgramRunner to measure each program
    callbacks : List[MeasureCallback]
        MeasureCallback to be called after each measure batch
    num_measured : int
        Number of samples that have been measured
    best_time_cost : float
        The best running time (the smaller the better)
    best_index : int
        The index of the samples that the best schedule is in
    best_sch : Optional[Schedule]
        The best schedule found so far
    """

    builder: ProgramBuilder
    runner: ProgramRunner
    callbacks: List[MeasureCallback]
    num_measured: int
    best_time_cost: float
    best_index: int
    best_sch: Optional[Schedule]

    def __init__(
        self,
        builder: Union[str, ProgramBuilder] = "local",
        runner: Union[str, ProgramRunner] = "rpc",
        measure_callbacks: Optional[List[MeasureCallback]] = None,
    ):
        if not isinstance(builder, ProgramBuilder):
            builder = ProgramBuilder.create(builder)
        if not isinstance(runner, ProgramRunner):
            runner = ProgramRunner.create(runner)
        if measure_callbacks is None:
            measure_callbacks = []
        self.__init_handle_by_constructor__(
            _ffi_api.ProgramMeasurer,  # pylint: disable=no-member
            builder,
            runner,
            measure_callbacks,
        )


########## ProgramTester ##########


class ProgramTester:
    """A utility function to run a specific PrimFunc on RPC"""

    target: Target
    target_host: Optional[Target]
    build_func: str
    rpc_key: str
    rpc_host: str
    rpc_port: int

    def __init__(
        self,
        target: Target,
        target_host: Optional[Target] = None,
        build_func: str = "tar",
        rpc_key: Optional[str] = None,
        rpc_host: Optional[str] = None,
        rpc_port: Optional[int] = None,
    ):
        super().__init__()
        if rpc_key is None:
            rpc_key = os.environ.get("TVM_TRACKER_KEY", None)
            if rpc_key is None:
                raise ValueError(
                    "RPC device key is not provided. Please provide 'rpc_key' explicitly, "
                    "or set environment variable TVM_TRACKER_KEY"
                )
        if rpc_host is None:
            rpc_host = os.environ.get("TVM_TRACKER_HOST", None)
            if rpc_host is None:
                raise ValueError(
                    "RPC tracker's host address is not provided. Please provide "
                    "'rpc_host' explicitly, or set environment variable TVM_TRACKER_HOST"
                )
        if rpc_port is None:
            rpc_port = os.environ.get("TVM_TRACKER_PORT", None)
            if rpc_port is None:
                raise ValueError(
                    "RPC tracker's host address is not provided. Please provide 'port' explicitly, "
                    "or set environment variable TVM_TRACKER_PORT"
                )
            rpc_port = int(rpc_port)

        self.target = target
        self.target_host = target_host
        self.build_func = build_func
        self.rpc_key = rpc_key
        self.rpc_host = rpc_host
        self.rpc_port = rpc_port

    def __call__(self, func: PrimFunc, args: List[np.ndarray]):
        """Build and run the PrimFunc with the specific arguments

        Parameters
        ----------
        func : PrimFunc
            The TIR func to be built and run
        args : List[np.ndarray]
            The list of arguments to run

        Returns
        -------
        filename : str
            The path to the exported library
        """
        filename = self._build_prim_func(func)
        args = self._run_exported_library(filename, args)
        return args

    def _build_prim_func(self, func: PrimFunc) -> str:
        """Build a PrimFunc

        Parameters
        ----------
        func : PrimFunc
            The TIR func to be built

        Returns
        -------
        filename : str
            The path to the exported library
        """
        # pylint: disable=import-outside-toplevel
        import tempfile

        from tvm.driver import build as tvm_build

        from ..autotvm.measure.measure_methods import set_cuda_target_arch
        from ..contrib import ndk as build_func_ndk
        from ..contrib import tar as build_func_tar

        # pylint: enable=import-outside-toplevel

        build_func = self.build_func
        build_func = {
            "tar": build_func_tar.tar,  # export to tar
            "ndk": build_func_ndk.create_shared,  # export to ndk
        }.get(build_func, build_func)
        if isinstance(build_func, str):
            raise ValueError("Invalid build_func: " + build_func)
        if self.target.kind.name == "cuda":
            set_cuda_target_arch(self.target.attrs["arch"])
        func = tvm_build(
            func,
            target=self.target,
            target_host=self.target_host,
        )
        filename = os.path.join(tempfile.mkdtemp(), "tmp_func." + build_func.output_format)
        func.export_library(filename, build_func)
        return filename

    def _run_exported_library(
        self,
        filename: str,
        args: List[np.ndarray],
    ) -> List[np.ndarray]:
        """Run an exported library

        Parameters
        ----------
        filename : str
            The exported library to be uploaded
        args : List[np.ndarray]
            The list of arguments to run

        Returns
        -------
        args : List[np.ndarray]
            The result after executing the exported library
        """

        from .utils import request_remote  # pylint: disable=import-outside-toplevel

        # upload built module
        _, remote = request_remote(self.rpc_key, self.rpc_host, self.rpc_port)
        remote.upload(filename)
        func = remote.load_module(os.path.split(filename)[1])
        ctx = remote.context(dev_type=self.target.kind.name, dev_id=0)
        args = [ndarray.array(arg, ctx=ctx) for arg in args]
        func(*args)
        args = [arg.asnumpy() for arg in args]
        return args

File: src/meta_schedule/measure.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "./measure.h"  // NOLINT(build/include)

#include <tvm/node/serialization.h>

#include <algorithm>

#include "./utils.h"

namespace tvm {
namespace meta_schedule {

/********** LocalBuilder: ProgramBuilder **********/

/*! \brief LocalBuilder use local CPU cores to build programs in parallel */
class LocalBuilderNode : public ProgramBuilderNode {
 public:
  /*! \brief Build function, can be `tar` or `ndk`. */
  String build_func;

  void VisitAttrs(tvm::AttrVisitor* v) {
    v->Visit("n_parallel", &n_parallel);
    v->Visit("timeout", &timeout);
    v->Visit("build_func", &build_func);
  }

  /*! \brief Default destructor */
  ~LocalBuilderNode() = default;

  Array<BuildResult> Build(const Array<MeasureInput>& inputs, int verbose) const override {
    if (const auto* f = runtime::Registry::Get("meta_schedule.local_builder.build")) {
      Array<BuildResult> results = (*f)(inputs, timeout, n_parallel, build_func, verbose);
      return results;
    }
    LOG(FATAL) << "meta_schedule.local_builder.build is not registered. "
               << "This is a function registered in Python, "
               << "make sure the TVM Python runtime has been loaded successfully.";
    throw;
  }

  static constexpr const char* _type_key = "meta_schedule.LocalBuilder";
  TVM_DECLARE_FINAL_OBJECT_INFO(LocalBuilderNode, ProgramBuilderNode);
};

/*!
 * \brief Managed reference to LocalBuilderNode.
 * \sa LocalBuilderNode
 */
class LocalBuilder : public ProgramBuilder {
 public:
  /*!
   * \brief The constructor.
   * \param timeout The timeout limit (in second) for each build process.
   * This will be used in a wrapper of the multiprocessing.Process.join().
   * \param n_parallel The number of threads used to build in parallel.
   * \param build_func The name of the registered build function.
   */
  explicit LocalBuilder(int timeout, int n_parallel, String build_func) {
    if (build_func != "tar" && build_func != "ndk") {
      LOG(FATAL) << "ValueError: Unknown build_func in LocalBuilder: " << build_func;
    }
    ObjectPtr<LocalBuilderNode> n = make_object<LocalBuilderNode>();
    n->timeout = timeout;
    n->n_parallel = n_parallel;
    n->build_func = std::move(build_func);
    data_ = std::move(n);
  }

  TVM_DEFINE_MUTABLE_OBJECT_REF_METHODS(LocalBuilder, ProgramBuilder, LocalBuilderNode);
};

/********** RPCRunner: ProgramRunner **********/

/*!
 * \brief RPCRunner that uses RPC call to measures the time cost of programs on remote devices.
 * Or sometime we may need to use RPC even in local running to insulate the thread environment.
 * (e.g. running CUDA programs)
 */
class RPCRunnerNode : public ProgramRunnerNode {
 public:
  /*! \brief The key of the device registered in the RPC tracker. */
  String key;
  /*! \brief The host address of the RPC tracker. */
  String host;
  /*! \brief The port of RPC tracker */
  int port;
  /*! \brief The priority of this run request, larger is more prior. */
  int priority;
  /*! \brief The number of tasks run in parallel. */
  int n_parallel;

  void VisitAttrs(tvm::AttrVisitor* v) {
    v->Visit("timeout", &timeout);
    v->Visit("number", &number);
    v->Visit("repeat", &repeat);
    v->Visit("min_repeat_ms", &min_repeat_ms);
    v->Visit("cooldown_interval", &cooldown_interval);
    v->Visit("enable_cpu_cache_flush", &enable_cpu_cache_flush);
    v->Visit("key", &key);
    v->Visit("host", &host);
    v->Visit("port", &port);
    v->Visit("priority", &priority);
    v->Visit("n_parallel", &n_parallel);
  }

  /*! \brief Default destructor */
  ~RPCRunnerNode() = default;

  Array<MeasureResult> Run(const Array<MeasureInput>& inputs,
                           const Array<BuildResult>& build_results, int verbose) const override {
    if (const auto* f = runtime::Registry::Get("meta_schedule.rpc_runner.run")) {
      Array<MeasureResult> results = (*f)(
          inputs, build_results, key, host, port, priority, n_parallel, timeout, number, repeat,
          min_repeat_ms, cooldown_interval, enable_cpu_cache_flush, f_create_args, verbose);
      return results;
    }
    LOG(FATAL) << "meta_schedule.rpc_runner.run is not registered. "
               << "This is a function registered in Python, "
               << "make sure the TVM Python runtime has been loaded successfully.";
    throw;
  }

  static constexpr const char* _type_key = "meta_schedule.RPCRunner";
  TVM_DECLARE_FINAL_OBJECT_INFO(RPCRunnerNode, ProgramRunnerNode);
};

/*!
 * \brief Managed reference to RPCRunnerNode.
 * \sa RPCRunnerNode
 */
class RPCRunner : public ProgramRunner {
 public:
  /*!
   * \brief The constructor. See the corresponding class in python/tvm/meta_schedule/measure.py
   * for more detailed parameter explanation.
   * \param key The key of the device registered in the RPC tracker.
   * \param host The host address of the RPC Tracker.
   * \param port The port of RPC Tracker.
   * \param priority The priority of this run request, larger is more prior.
   * \param n_parallel The number of tasks run in parallel.
   * \param timeout Timeout of a run.
   * \param number The number of times to run the generated code for taking average.
   * \param repeat The number of times to repeat the measurement.
   * \param min_repeat_ms The minimum duration of one repeat in milliseconds.
   * \param cooldown_interval The cool down interval between two measurements.
   * \param enable_cpu_cache_flush Whether to flush cache on CPU between repeated measurements.
   * \param f_create_args Callback function to create arguments.
   */
  explicit RPCRunner(String key, String host, int port, int priority, int n_parallel, int timeout,
                     int number, int repeat, int min_repeat_ms, double cooldown_interval,
                     bool enable_cpu_cache_flush, FCreateArgs f_create_args) {
    ObjectPtr<RPCRunnerNode> n = make_object<RPCRunnerNode>();
    n->key = std::move(key);
    n->host = std::move(host);
    n->port = port;
    n->priority = priority;
    n->timeout = timeout;
    n->n_parallel = n_parallel;
    n->number = number;
    n->repeat = repeat;
    n->min_repeat_ms = min_repeat_ms;
    n->cooldown_interval = cooldown_interval;
    n->enable_cpu_cache_flush = enable_cpu_cache_flush;
    n->f_create_args = f_create_args;
    data_ = std::move(n);
  }

  TVM_DEFINE_MUTABLE_OBJECT_REF_METHODS(RPCRunner, ProgramRunner, RPCRunnerNode);
};

/********** ProgramMeasurer **********/

ProgramMeasurer::ProgramMeasurer(ProgramBuilder builder, ProgramRunner runner,
                                 Array<MeasureCallback> callbacks, Database db, int num_measures) {
  ObjectPtr<ProgramMeasurerNode> n = make_object<ProgramMeasurerNode>();
  n->builder = std::move(builder);
  n->runner = std::move(runner);
  n->callbacks = std::move(callbacks);
  n->db = std::move(db);
  n->num_measures = num_measures;
  data_ = std::move(n);
}

ProgramMeasurer::ProgramMeasurer(ProgramBuilder builder, ProgramRunner runner,
                                 Array<MeasureCallback> callbacks)
    : ProgramMeasurer(builder, runner, callbacks, InMemoryDB(NullOpt), /*num_measures=*/0) {}

void ProgramMeasurerNode::Init(const SearchTask& task) {
  for (const MeasureCallback& callback : callbacks) {
    callback->Init(task);
  }
  this->db = InMemoryDB(task->log_file);
  this->db->Init(task);
}

Optional<Schedule> ProgramMeasurerNode::GetBest(const SearchTask& task) const {
  Optional<Trace> trace = db->GetBest().trace;
  if (!trace.defined()) {
    return NullOpt;
  }
  Schedule sch(task->workload);
  trace.value()->Apply(sch);
  return sch;
}

Array<MeasureResult> ProgramMeasurerNode::PureMeasure(const Array<MeasureInput>& measure_inputs,
                                                      int verbose) const {
  Array<BuildResult> build_results = builder->Build(measure_inputs, verbose);
  Array<MeasureResult> measure_results = runner->Run(measure_inputs, build_results, verbose);
  return measure_results;
}

Array<MeasureResult> ProgramMeasurerNode::BatchMeasure(const Array<MeasureInput>& measure_inputs,
                                                       int batch_size, int verbose) {
  Array<MeasureResult> measure_results;
  int n_samples = measure_inputs.size();
  for (int st = 0; st < n_samples; st += batch_size) {
    int ed = std::min(st + batch_size, n_samples);
    Array<MeasureInput> batch_measure_inputs(measure_inputs.begin() + st,
                                             measure_inputs.begin() + ed);
    Array<MeasureResult> batch_measure_results = this->PureMeasure(batch_measure_inputs, verbose);
    for (int i = 0; i < ed - st; ++i) {
      const MeasureInput& measure_input = batch_measure_inputs[i];
      const MeasureResult& measure_result = batch_measure_results[i];
      double flop_ct = measure_input->task->flop_ct;
      const String& task_name = measure_input->task->task_name;
      MeasureErrorNO error_no = static_cast<MeasureErrorNO>(measure_result->error_no);
      ++num_measures;
      if (error_no == MeasureErrorNO::kNoError) {
        double avg_time_cost = FloatArrayMean(measure_result->costs);
        db->Add(measure_input->sch->trace, Repr(measure_input->sch),
                AsVector<FloatImm, double>(measure_result->costs));
        double best_time_cost = db->GetBest().MeanTime();
        StdCout(verbose) << std::fixed << std::setprecision(4)  //
                         << '[' << task_name << "] #" << num_measures
                         << "\tTime: " << (avg_time_cost * 1000) << " ms, "
                         << (flop_ct / avg_time_cost / 1e9) << " GFLOPs"
                         << "\tBest time: " << (best_time_cost * 1000) << " ms, "
                         << (flop_ct / best_time_cost / 1e9) << " GFLOPs" << std::endl;
      } else if (error_no == MeasureErrorNO::kRunTimeoutError ||
                 error_no == MeasureErrorNO::kBuildTimeoutError) {
        double best_time_cost = db->GetBest().MeanTime();
        StdCout(verbose) << std::fixed << std::setprecision(4)  //
                         << '[' << task_name << "] #" << num_measures
                         << "\tError: " << MeasureErrorNOToStr(error_no)
                         << "\tBest time: " << (best_time_cost * 1000) << " ms, "
                         << (flop_ct / best_time_cost / 1e9) << " GFLOPs" << std::endl;
      } else {
        double best_time_cost = db->GetBest().MeanTime();
        StdCout(verbose) << std::fixed << std::setprecision(4)  //
                         << '[' << task_name << "] #" << num_measures
                         << "\tError: " << MeasureErrorNOToStr(error_no)
                         << "\tBest time: " << (best_time_cost * 1000) << " ms, "
                         << (flop_ct / best_time_cost / 1e9) << " GFLOPs" << std::endl
                         << measure_result->error_msg << "\n"
                         << "The IR is:\n"
                         << Repr(measure_input->sch) << "\nSchedule is:\n"
                         << measure_input->sch->trace->Stringify();
      }
    }
    measure_results.insert(measure_results.end(), batch_measure_results.begin(),
                           batch_measure_results.end());
    for (const MeasureCallback& callback : this->callbacks) {
      callback->Callback(measure_inputs, measure_results);
    }
  }
  return measure_results;
}

/********** FFI **********/

struct Internal {
  /********** Constructors **********/
  /*!
   * \brief The constructor.
   * \param timeout The timeout limit (in second) for each build process.
   * \param n_parallel The number of threads used to build in parallel.
   * \param build_func The name of the registered build function.
   * \return The LocalBuilder constructed
   * \sa LocalBuilder::LocalBuilder
   */
  static LocalBuilder LocalBuilderNew(int timeout, int n_parallel, String build_func) {
    return LocalBuilder(timeout, n_parallel, build_func);
  }
  /*!
   * \brief Constructor of RPCRunner
   * \param key The key of the device registered in the RPC tracker.
   * \param host The host address of the RPC Tracker.
   * \param port The port of RPC Tracker.
   * \param priority The priority of this run request, larger is more prior.
   * \param n_parallel The number of tasks run in parallel.
   * \param timeout Timeout of a run.
   * \param number The number of times to run the generated code for taking average.
   * \param repeat The number of times to repeat the measurement.
   * \param min_repeat_ms The minimum duration of one repeat in milliseconds.
   * \param cooldown_interval The cool down interval between two measurements.
   * \param enable_cpu_cache_flush Whether to flush cache on CPU between repeated measurements.
   * \return The RPCRunner constructed
   * \sa RPCRunnerNew::RPCRunnerNew
   */
  static RPCRunner RPCRunnerNew(String key, String host, int port, int priority, int n_parallel,
                                int timeout, int number, int repeat, int min_repeat_ms,
                                double cooldown_interval, bool enable_cpu_cache_flush,
                                ProgramRunner::FCreateArgs f_create_args) {
    return RPCRunner(key, host, port, priority, n_parallel, timeout, number, repeat, min_repeat_ms,
                     cooldown_interval, enable_cpu_cache_flush, f_create_args);
  }
  /*!
   * \brief Constructor of ProgramMeasurerNew
   * \param builder The program builder
   * \param runner The program runner
   * \param callbacks The callbacks invoked after measurement
   * \return The ProgramMeasurer constructed
   * \sa ProgramMeasurer::ProgramMeasurer
   */
  static ProgramMeasurer ProgramMeasurerNew(ProgramBuilder builder, ProgramRunner runner,
                                            Array<MeasureCallback> callbacks) {
    return ProgramMeasurer(builder, runner, callbacks);
  }
  /********** Member methods **********/
  /*!
   * \brief Invoke ProgramBuilder::Build
   * \param builder The program builder
   * \param inputs An Array of MeasureInput
   * \param verbose Verbosity level. 0 for silent, 1 to output information during program building
   * \return An Array of MeasureResult
   * \sa ProgramBuilder::Build
   */
  static Array<BuildResult> ProgramBuilderBuild(ProgramBuilder builder, Array<MeasureInput> inputs,
                                                int verbose) {
    return builder->Build(inputs, verbose);
  }
  /*!
   * \brief Invoke ProgramRunner::Run
   * \param runner The program runner
   * \param inputs An Array of MeasureInput
   * \param build_results An Array of BuildResult
   * \param verbose Verbosity level. 0 for silent, 1 to output information during program running
   * \return An Array of MeasureResult
   * \sa ProgramRunner::Run
   */
  static Array<MeasureResult> ProgramRunnerRun(ProgramRunner runner, Array<MeasureInput> inputs,
                                               Array<BuildResult> build_results, int verbose) {
    return runner->Run(inputs, build_results, verbose);
  }
};

TVM_REGISTER_OBJECT_TYPE(ProgramRunnerNode);
TVM_REGISTER_OBJECT_TYPE(ProgramBuilderNode);
TVM_REGISTER_NODE_TYPE(LocalBuilderNode);
TVM_REGISTER_NODE_TYPE(RPCRunnerNode);
TVM_REGISTER_OBJECT_TYPE(MeasureCallbackNode);
TVM_REGISTER_NODE_TYPE(ProgramMeasurerNode);

TVM_REGISTER_GLOBAL("meta_schedule.LocalBuilder").set_body_typed(Internal::LocalBuilderNew);
TVM_REGISTER_GLOBAL("meta_schedule.RPCRunner").set_body_typed(Internal::RPCRunnerNew);
TVM_REGISTER_GLOBAL("meta_schedule.ProgramMeasurer").set_body_typed(Internal::ProgramMeasurerNew);
TVM_REGISTER_GLOBAL("meta_schedule.ProgramBuilderBuild")
    .set_body_typed(Internal::ProgramBuilderBuild);
TVM_REGISTER_GLOBAL("meta_schedule.ProgramRunnerRun").set_body_typed(Internal::ProgramRunnerRun);

}  // namespace meta_schedule
}  // namespace tvm
